github/release_notes/__init__.py

```diff
@@ -0,0 +1,353 @@
+        )) \
+        .value()
+    return commits
+
+def fetch_pr_numbers_from_commits(
+    commits: list
+) -> set:
+    pr_numbers = set()
+    for commit in commits:
+        pr_number = pr_number_from_subject(commit.subject)
+
+        if pr_number:
+            pr_numbers.add(pr_number)
+
+    return pr_numbers
+
+def pr_number_from_subject(commit_subject: str):
+    pr_number = _.head(re.findall(r"Merge pull request #(\d+)", commit_subject))
+    if not pr_number: # Squash commit
+        pr_number = _.head(re.findall(r" \(#(\d+)\)$", commit_subject))
+    return pr_number
+
+def fetch_release_notes_from_prs(
+    helper: GitHubRepositoryHelper,
+    pr_numbers_in_range: set,
+    cn_current_repo: ComponentName
+) -> list:
+    # we should consider adding a release-note label to the PRs
+    # to reduce the number of search results
+    prs_iter = helper.search_issues_in_repo('type:pull is:closed')
+
+    release_notes = list()
+    for pr_iter in prs_iter:
+        pr_dict = pr_iter.as_dict()
+
+        pr_number = pr_dict['number']
+        if not str(pr_number) in pr_numbers_in_range:
+            continue
+
+        release_notes_pr = extract_release_notes(
+            reference_id=pr_number,
+            text=pr_dict['body'],
+            user_login=_.get(pr_dict, 'user.login'),
+            cn_current_repo=cn_current_repo,
+            reference_is_pr=True
+        )
+        if not release_notes_pr:
+            continue
+
+        release_notes.extend(release_notes_pr)
+    return release_notes
+
+def fetch_release_notes_from_commits(
+    commits: list,
+    cn_current_repo: ComponentName
+):
+    release_notes = list()
+    for commit in commits:
+
+        release_notes_commit = extract_release_notes(
+            reference_id=commit.hash,
+            text=commit.message,
+            user_login=None, # we do not have the gitHub user at hand
+            cn_current_repo=cn_current_repo,
+            reference_is_pr=False
+        )
+        if not release_notes_commit:
+            continue
+
+        release_notes.extend(release_notes_commit)
+    return release_notes
+
+def extract_release_notes(
+    reference_is_pr: bool,
+    reference_id: int,
+    text: str,
+    user_login: str,
+    cn_current_repo: ComponentName,
+) -> list:
+    release_notes = list()
+
+    r = re.compile(
+        r"``` *(?P<category>improvement|noteworthy) (?P<target_group>user|operator)"
+        "( (?P<source_repo>\S+/\S+/\S+)(( (?P<reference_type>#|\$)(?P<reference_id>\S+))?"
+        "( @(?P<user>\S+))?)( .*?)?|( .*?)?)\r?\n(?P<text>.*?)\n```",
+        re.MULTILINE | re.DOTALL
+    )
+    for m in r.finditer(text):
+        code_block = m.groupdict()
+
+        text = _.trim(code_block['text'])
+        if not text or 'none' == text.lower():
+            continue
+
+        category = code_block['category']
+        target_group = code_block['target_group']
+        source_repo = code_block['source_repo']
+        if source_repo:
```

Comment:
not sure what you mean. The source_repo is taken from the code block if there is any, if not, the current repository is set as source repository
If there is a source repository, it will also grab the reference id and user information and will not use the respective values passed as arguments